native-plasma/app/src/main/jni/plasma_renderer_rgb565.cpp

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright (C) 2015 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+#include "plasma_renderer.h"
+
+// Set to 1 to optimize memory stores when generating plasma.
+#define OPTIMIZE_WRITES 1
+
+/* Color palette used for rendering the plasma */
+static const int PALETTE_BITS = 8;
+static const int PALETTE_SIZE = 1 << PALETTE_BITS;
+
+static uint16_t palette[PALETTE_SIZE];
+
+static uint16_t makeColor(int red, int green, int blue) {
+  return (uint16_t) (((red << 8) & 0xf800) |
+                     ((green << 2) & 0x03e0) |
+                     ((blue >> 3) & 0x001f));
+}
+
+void PlasmaRendererRGB565::init_palette() {
```

Comment:
It could totally go in the parent class PlasmaRenderer as it's shared by
both derived class, what do you think ?

On Fri, 11 Dec 2015 at 21:36 Johan Euphrosine <notifications@github.com>
wrote:

> In native-plasma/app/src/main/jni/plasma_renderer_rgb565.cpp
> <https://github.com/googlesamples/android-ndk/pull/133#discussion_r47408221>
> :
>
> > +// Set to 1 to optimize memory stores when generating plasma.
> > +#define OPTIMIZE_WRITES 1
> > +
> > +/* Color palette used for rendering the plasma */
> > +static const int PALETTE_BITS = 8;
> > +static const int PALETTE_SIZE = 1 << PALETTE_BITS;
> > +
> > +static uint16_t palette[PALETTE_SIZE];
> > +
> > +static uint16_t makeColor(int red, int green, int blue) {
> > +  return (uint16_t) (((red << 8) & 0xf800) |
> > +                     ((green << 2) & 0x03e0) |
> > +                     ((blue >> 3) & 0x001f));
> > +}
> > +
> > +void PlasmaRendererRGB565::init_palette() {
>
> Maybe you could share that function and take the palette size as an arg,
> or maybe it a template parameter?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/googlesamples/android-ndk/pull/133/files#r47408221>.
>